
syntax = "proto3";

package SafeEntry;

// Note: Come back to
service TrackingService  {

    rpc Login (User) returns (User) {}

    rpc GetAllLocations(Empty) returns (stream Location) {}

    rpc CreateCheckInIndividual(Location) returns (Status) {}
    rpc CreateCheckInGroup(Location) returns (Status) {}
    rpc CreateCheckOutIndividual(CheckOut) returns (Status) {}
    rpc CreateCheckOutGroup(CheckOut) returns (Status) {}

    rpc GetCheckOutOptionsIndividual(User) returns (stream CheckOut) {}
    rpc GetCheckOutOptionsGroup(User) returns (stream CheckOut) {}

    rpc GetAllUsers(User) returns (stream User) {}
    rpc GetGroupsByUser(User) returns (stream Group) {}
    rpc CreateGroup(Group) returns (Group) {}
    rpc AddUserToGroup(stream User) returns (Status) {}
    
    rpc GetSafeEntry(User) returns (stream SafeEntry) {}
    
    rpc GetCovidLocationByUser (User) returns (stream Location) {}
    rpc CreateReportCovidCase (stream Location) returns (Status) {}

    rpc GetAllNotificiationByUser (User) returns (stream Notificiation) {}
}

message Empty {

}

message User {
    string name = 1;
    string nric = 2;
    string role_name = 3;
    Group group = 4;
    Status status = 5;
}

message Group {
    string name = 1;
    User user = 2;
    Status status = 3;
}

message Status {
    bool status = 1;
}

message Location {
    User user = 1;
    int32 id = 2;
    Group group = 3;
    string name = 4;
}

message CheckOut {
    int32 id = 1;
    Location location = 2;
    string check_in = 3;
    Group group = 4;
}

message SafeEntry {
    Location location = 1;
    Group group = 2;
    string checkin = 3;
    string checkout = 4;
}

message Notificiation {
    Location location = 1;
    string checkin = 2;
    string case_date = 3;
}

message Case {
    User user = 1;
    Location location = 2;
}

//python3 -m grpc_tools.protoc -I . --python_out=. --grpc_python_out=. ./tracking.proto